### PR TITLE
Plugin system

### PR DESCRIPTION
**Pull request by jcgertig**
The base for a plugin system and I build one plugin to test with

### CODE REVIEW

**Comment by jcgertig** [comment]
My screen record tool is cutting short so here are 3 gifs to show the features of the plugin.

one

two

three

**Comment by nitin42** [comment]
Looks awesome!! A plugin system for this component ? Right ?

**Comment by jcgertig** [comment]
Yes I added a plugin system for this component

**Comment by nitin42** [comment]
Sorry I accidentally clicked on close PR button. `terminal-in-react-pseudo-file-system-plugin` seems like you're developing a plugin system where any user can create a plugin for the component. Awesome!!

**Comment by jcgertig** [comment]
The gifs are of the plugin I created here [https://github.com/jcgertig/terminal-in-react-pseudo-file-system-plugin](https://github.com/jcgertig/terminal-in-react-pseudo-file-system-plugin)

**Comment by jcgertig** [comment]
the code in this pr allows anyone to add plugins

**Comment by jcgertig** [comment]
`terminal-in-react-pseudo-file-system-plugin` is just my test plugin

**Comment by nitin42** [comment]
I want to discuss something with you. But where ?

**Comment by jcgertig** [comment]
The changes in this pr are breaking so this would have to be a major release

**Comment by jcgertig** [comment]
add a gitter chat for the repo?

**Comment by nitin42** [comment]
Ok! I will go ahead and make a chat room.

**Comment by nitin42** [comment]
Can you resolve the conflicts ? I've pushed some changes (merged history feature) !

**Comment by jcgertig** [comment]
@nitin42 done also im looking into the hound ci issue

**Comment by nitin42** [comment]
Did you also fixed this👇  in this PR ?

**Comment by jcgertig** [comment]
No in the last one

**Comment by nitin42** [comment]
One more thing, last time I checked color, backgroundColor and barColor were properly applied. Facing some gotchas now!

**Comment by nitin42** [comment]
Try changing these props... and let me know if its working or not!

**Comment by nitin42** [comment]
I think we are overriding some of the styles. May be that's why ?

**Comment by nitin42** [comment]
See this :

**Comment by jcgertig** [comment]
@nitin42 fixed

**Comment by jcgertig** [comment]
Think we can go ahead and add this as the v1 of the plugin system Im still working on new bits but will be a while

**Comment by jcgertig** [comment]
Just remember these changes are breaking to older versions

**Comment by jcgertig** [comment]
@nitin42 ^

**Comment by nitin42** [comment]
Looks great! I'm going to my bed now but I'll merge this in the morning and will publish the V1 then 😄

File: components/Terminal.js [hunk 1]
```diff
@@ -33,6 +34,23 @@ function handleLogging(method, addToOutput) {
...
+      description: PropTypes.string,
+      defaultValue: PropTypes.any,
+    })),
+    method: PropTypes.func,
+  }),
+]));
+
+const descriptionsPropType = PropTypes.objectOf(PropTypes.oneOfType([
+  PropTypes.string,
+  PropTypes.bool,
```
Comment (nitin42) [inline components/Terminal.js]: This hides built-in options. Right ?

File: components/Terminal.js [hunk 2]
```diff
@@ -43,24 +61,20 @@ class Terminal extends Component {
...
-    description: PropTypes.objectOf(PropTypes.string),
+    commands: commandsPropType,
+    descriptions: descriptionsPropType,
     watchConsoleLogging: PropTypes.bool,
     commandPassThrough: PropTypes.oneOfType([
       PropTypes.func,
       PropTypes.bool,
     ]),
     promptSymbol: PropTypes.string,
+    plugins: PropTypes.arrayOf(PropTypes.shape({
```
Comment (nitin42) [inline components/Terminal.js]: Its great that we have a plugin system but what about the docs ? about creating a one ?

File: components/Terminal.js [hunk 3]
```diff
@@ -330,7 +330,7 @@ class Terminal extends Component {
     let res;
 
     if (input === '') {
-      this.printLine('');
+      // do nothing
```
Comment (nitin42) [inline components/Terminal.js]: Thanks!

**Comment by nitin42** [comment]
Close, Minimise and Maximise not working !! Please look into this..

**Comment by nitin42** [comment]
Yup. Checked everything! not working 👇

**Comment by nitin42** [comment]
Close, Max and Min were not working but I've fixed this. Thanks!

**Pull request merged**